Yoast/wordpress-seo
# Implement Roles and Capabilities

## Summary

This PR can be summarized in the following changelog entry:

* Introduces `wpseo_manager` and `wpseo_editor` roles.
* Introduces capabilities to control which user has access to the advanced SEO settings.

### Touches

* All SEO settings pages
* Network Admin SEO Settings
* SEO notifications
* Page/post/term edit pages (Advanced settings)
* Reindex internal links

## Relevant technical choices:

* Used abstract classes to reduce duplicating code
* Separated registering and adding/removing
* Introduced actions to register roles and capabilities on, making it extensible for premium
* `wpseo_manage_options` can do anything, thus this is always being added to the capability check
* menu registration is done the regular way when the `wpseo_manage_options` capability is present
* submenus with capabilities other than `wpseo_manage_options` are registered as main menu items when the used does not have `wpseo_manage_options`

## Test instructions

Recommended helper plugins:
* Members
* User switching

This PR can be tested by following these steps:

* Deactivate + activate the plugin to see the new roles and capabilities being added to the system
* Create a user with a specific role (wpseo_manager / wpseo_editor) and verify that that user can only do the things they should be able to do.

Notes:
* The `Advanced SEO Settings` in a post or page needs to be disabled in `General > Security` in order to hide it for users who don't have the proper capability.
* In a Network setup, we have a setting which disables `Administrators` from being able to modify SEO settings, if this is configured, we don't give this role the capability `wpseo_manage_options`

Fixes #7831
